datastream-kafka/src/main/java/com/linkedin/datastream/kafka/KafkaTransportProviderAdmin.java

```diff
@@ -245,47 +243,90 @@ public void dropDestination(Datastream datastream) {
-    Properties props = AdminUtils.fetchEntityConfig(_zkUtils.get(), ConfigType.Topic(), topicName);
-    if (!props.containsKey(TOPIC_RETENTION_MS)) {
-      return null;
+
+    try {
+      ConfigResource topicResource = new ConfigResource(ConfigResource.Type.TOPIC, topicName);
+      Map<ConfigResource, Config> topicConfigMap = adminClient.describeConfigs(Collections.singletonList(topicResource)).all().get();
+      Config topicConfig = topicConfigMap.get(topicResource);
+      ConfigEntry entry = topicConfig.get(TOPIC_RETENTION_MS);
+      if (entry != null) {
+        return Duration.ofMillis(Long.parseLong(entry.value()));
+      }
+    } catch (InterruptedException | ExecutionException e) {
+      LOG.warn("Failed to retrieve config for topic {}.", topicName, e);
     }
-    return Duration.ofMillis(Long.parseLong(props.getProperty(TOPIC_RETENTION_MS)));
+
+    return null;
   }
 
   /**
    * Create Kafka topic based on the destination connection string, if it does not already exist.
    * @param connectionString connection string from which to obtain topic name
    * @param numberOfPartitions number of partitions
-   * @param topicConfig topic config to use for topic creation
+   * @param topicProperties topic config to use for topic creation
    */
-  public void createTopic(String connectionString, int numberOfPartitions, Properties topicConfig) {
+  public void createTopic(String connectionString, int numberOfPartitions, Properties topicProperties, Datastream datastream) {
     Validate.notNull(connectionString, "destination should not be null");
-    Validate.notNull(topicConfig, "topicConfig should not be null");
-    Validate.isTrue(_zkUtils.isPresent(), "zkUtils should be present");
+    Validate.notNull(topicProperties, "topicConfig should not be null");
 
     String topicName = KafkaTransportProviderUtils.getTopicName(connectionString);
-    populateTopicConfig(topicConfig);
-    try {
-      // Create only if it doesn't exist.
-      if (!AdminUtils.topicExists(_zkUtils.get(), topicName)) {
-        int replicationFactor = Integer.parseInt(topicConfig.getProperty("replicationFactor", DEFAULT_REPLICATION_FACTOR));
-        LOG.info("Creating topic with name {} partitions={} with properties {}", topicName, numberOfPartitions,
-                topicConfig);
+    populateTopicConfig(topicProperties);
 
-        AdminUtils.createTopic(_zkUtils.get(), topicName, numberOfPartitions, replicationFactor, topicConfig, RackAwareMode.Disabled$.MODULE$);
-      } else {
+    try {
+      AdminClient adminClient = getAdminClient(datastream);
+      NewTopic newTopic = getTopic(topicName, numberOfPartitions, topicProperties);
+      // Removing from topic config since its passed as a direct argument
+      topicProperties.remove("replicationFactor");
+      // TopicExistsException is thrown if topic exists. Its a no-op.
+      adminClient.createTopics(Collections.singletonList(newTopic)).all().get();
```

Comment:
By doing ".all().get()" we have a blocking wait for the topic to be created. I am not sure what exactly was the behavior with AdminUtils and/or if doing a blocking wait is the correct thing to do here. I am guessing this is the correct way?

However, having said that. On repeated execution of all the unit tests, I noticed that the test "testCreateDatastreamHappyPathDefaultRetention" is a little flaky in that it randomly fails every now and then (very rarely though, hard to reproduce). This test calls "getRetention" which queries topic config and returns retention time if its exists in policy else null, and when this test fails its because getRetention returned null instead of actual expected value. I am not sure if this test was flaky even before. I am not sure how to explain this behavior.

Reply:
Summarizing the offline discussion here:

There's no callbacks and returned futures with `AdminUtils.createTopic`. No hints that it's an asynchronous operation. I'm not a scala expert, but from the looks of it it's a synchronous call.

As for the failing/flaky test:

1. Was this test flaky even before or it became flaky as a result of the API changes? We need to rule out changes in behavior of the system after switching from `AdminUtils` to `AdminClient`
2. Does adding a thread sleep after topic creation (and before querying the retention of the new topic) make the test pass? 

For (1), we'll reach out to Kafka to confirm that there are no changes in behavior in createTopic APIs. For (2) we can try fixing the flaky test in a separate PR.